src/rust/engine/fs/store/src/snapshot_ops.rs

```diff
@@ -0,0 +1,731 @@
+                  {
+                    globbed_directories.insert(directory_path, directory_node.clone());
+                    continue;
+                  }
+                  _ => (),
+                }
+                let with_double_star = RestrictedPathGlob::DirWildcard {
+                  wildcard: wildcard.clone(),
+                  remainder: remainder.clone(),
+                };
+                subdir_globs.push(with_double_star);
+              } else {
+                assert_ne!(0, remainder.len());
+              }
+              match remainder.len() {
+                0 => (),
+                1 => {
+                  let next_glob = RestrictedPathGlob::Wildcard {
+                    wildcard: remainder.get(0).unwrap().clone(),
+                  };
+                  subdir_globs.push(next_glob);
+                }
+                _ => {
+                  let next_glob = RestrictedPathGlob::DirWildcard {
+                    wildcard: remainder.get(0).unwrap().clone(),
+                    remainder: remainder[1..].to_vec(),
+                  };
+                  subdir_globs.push(next_glob);
+                }
+              }
+              // Append to the existing globs, and collate at the end of this iteration.
+              let entry = todo_directories
+                .entry(directory_path)
+                .or_insert_with(Vec::new);
+              entry.extend(subdir_globs);
+            }
+          }
+        }
+      }
+
+      // 1c. Push any "dormant" contexts to traverse subdirs onto the dormant stack.
+      let dependencies: Vec<PathBuf> = {
+        let mut dependencies: Vec<PathBuf> = vec![];
+        // If we ever encounter a directory *with* a remainder, we have to ensure that that
+        // directory is *not* in `globbed_directories`, which are *not* subsetted at all.
+        // NB: Because our goal is to *union* all of the includes, if a directory is in
+        // `globbed_directories`, we can skip processing it for subsetting here!
+        for subdir_path in todo_directories.keys() {
+          if !globbed_directories.contains_key(subdir_path) {
+            dependencies.push(prefix.join(subdir_path));
+          }
+        }
+        dependencies
+      };
+
+      for (subdir_name, all_path_globs) in todo_directories.into_iter() {
+        let full_name = prefix.join(&subdir_name);
+        let bazel_digest = cur_dir_directories
+          .get(&subdir_name)
+          .unwrap()
+          .get_digest()
+          .clone();
+        let digest = to_pants_digest(bazel_digest);
+        let multiple_globs = MultipleGlobs {
+          include: all_path_globs,
+          exclude: exclude.clone(),
+        };
+        unexpanded_stack.insert(
+          full_name,
+          UnexpandedSubdirectoryContext {
+            digest,
+            multiple_globs,
+          },
+        );
+      }
+
+      // 1d. Push a base "activated" context for the current directory node, then push dormant
```

Comment:
Not clear what active means either (by contrast). Would consider expanding both of these comments to replace the lingo with "stuff to visit on this iteration" vs "next iteration", or whatever it's referring to.

Reply:
Fixed up!